genesis/utils/geom.py

```diff
@@ -838,6 +838,124 @@ def transform_pos_quat_by_trans_quat(pos, quat, t_trans, t_quat):
     return new_pos, new_quat
 
 
+def transform_by_R(pos, R):
+    """
+    Transforms 3D points by a 3x3 rotation matrix or a batch of matrices,
+    supporting both NumPy arrays and PyTorch tensors.
+
+    Parameters
+    ----------
+    pos: np.ndarray | torch.Tensor
+        A numpy array or torch tensor of 3D points. Can be a single point
+         (3,), a batch of points (B, 3), or a batched batch of points (B, N, 3).
+    T: np.ndarray | torch.Tensor
+        The 3x3 rotation matrix or a batch of B rotation
+        matrices of shape (B, 3, 3). Must be of the same type as `pos`.
+
+    Returns
+    -------
+        The transformed points in a shape corresponding to the input dimensions.
+    """
+    assert pos.shape[-1] == 3
+
+    dim_added = False
+    if R.ndim == 2:
+        R = R[None]
+        dim_added = True
+    if pos.ndim == 3:
+        new_pos = (R @ pos.swapaxes(-1, -2)).swapaxes(-1, -2) 
+    elif pos.ndim == 2:
+        new_pos = (R @ pos[:, :, None])[...,0]
+    elif pos.ndim == 1:
+        new_pos = (R @ pos[None, :, None])[...,0]
+        if dim_added:
+            new_pos = new_pos[0]
+    else:
+        gs.raise_exception(f"invalid input dim for pos: {pos.shape=}")
+    return new_pos
+
+
+def transform_by_T(pos, T):
+    """
+    Transforms 3D points by a 4x4 transformation matrix or a batch of matrices,
+    supporting both NumPy arrays and PyTorch tensors.
+
+    Parameters
+    ----------
+    pos: np.ndarray | torch.Tensor
+        A numpy array or torch tensor of 3D points. Can be a single point
+         (3,), a batch of points (B, 3), or a batched batch of points (B, N, 3).
+    T: np.ndarray | torch.Tensor
+        The 4x4 transformation matrix or a batch of B transformation
+        matrices of shape (B, 4, 4). Must be of the same type as `pos`.
+
+    Returns
+    -------
+        The transformed points in a shape corresponding to the input dimensions.
+    """
+    assert pos.shape[-1] == 3, "Input positions must have 3 dimensions"
+
+    if isinstance(pos, torch.Tensor) and isinstance(T, torch.Tensor):
```

Comment:
Could you try to factories both implementations in one, for the part that could be shared ? Typically

```python
        pos_hom_t = pos_hom.swapaxes(-1, -2)  # (..., N, 4) -> (..., 4, N)
        transformed_hom = T @ pos_hom_t
        transformed_hom = transformed_hom.swapaxes(-1, -2)[..., :3]

        if pos.ndim == 1:
            transformed_hom = transformed_hom.reshape(-1)
        elif pos.ndim == 2:
            transformed_hom = transformed_hom.reshape(-1, 3)
        return transformed_hom
 ```

Should work in both numpy and torch ?